reconcile/terraform_repo.py

```diff
@@ -0,0 +1,304 @@
+    In regards to deleting a Terraform Repo, when the delete flag is set to True, then
+    the state representation of this repo is also deleted even though the definition may
+    still exist in App Interface
+    """
+    created_and_updated = {**created, **updated}
+    try:
+        for cu_key, cu_val in created_and_updated.items():
+            if cu_val.delete is True:
+                state.rm(cu_key)
+            else:
+                state.add(cu_key, cu_val.json(), True)
+        for d_key in deleted.keys():
+            state.rm(d_key)
+    except KeyError:
+        pass
+
+
+def calculate_diff(
+    existing_state: list[TFRepo],
+    desired_state: list[TFRepo],
+    dry_run: bool,
+    state: Optional[State],
+) -> list[TFRepo]:
+    """Diffs existing and desired state as well as updates the state in S3 if this is not a dry-run operation"""
+    existing_map = map_repos(existing_state)
+    desired_map = map_repos(desired_state)
+
+    to_be_created = diff_missing(desired_map, existing_map)
+    to_be_updated = diff_changed(existing_map, desired_map)
+
+    # indicates repos which have had their definitions deleted from App Interface
+    # a pre-requisite to this step is setting the delete flag in the repo definition
+    to_be_deleted = diff_missing(existing_map, desired_map, True)
+
+    if not dry_run and state:
+        update_state(to_be_created, to_be_deleted, to_be_updated, state)
+
+    return merge_results(to_be_created, to_be_updated)
+
+
+@defer
+def run(
+    dry_run: bool = True,
+    print_to_file: Optional[str] = None,
+    defer: Optional[Callable] = None,
+) -> None:
+
+    gqlapi = gql.get_api()
+
+    state = init_state(integration=QONTRACT_INTEGRATION)
+    if defer:
+        defer(state.cleanup)
+
+    desired = get_repos(query_func=gqlapi.query)
+    existing = get_existing_state(state)
+
+    repo_diff = calculate_diff(existing, desired, dry_run, state)
+
+    # validate that each new/updated repo is pointing to a valid git ref and
+    # has config.tf in the project path
+    for repo in repo_diff:
+        if not repo.delete:
+            check_ref(repo.repository, repo.ref, repo.project_path)
+
+    action_plan = ActionPlan(dry_run=dry_run, repos=repo_diff)
+
+    if print_to_file:
+        try:
+            with open(print_to_file, "w") as output_file:
+                yaml.safe_dump(
+                    data=action_plan.dict(), stream=output_file, explicit_start=True
+                )
+        except FileNotFoundError:
+            raise ParameterError(
+                "Unable to write to specified 'print_to_file' location: {}".format(
+                    print_to_file
+                )
+            )
+    else:
+        print(yaml.safe_dump(data=action_plan.dict(), explicit_start=True))
+
+
+def early_exit_desired_state(*args: Any, **kwargs: Any) -> dict[str, Any]:
+    gqlapi = gql.get_api()
+    return {"repos": get_repos(query_func=gqlapi.query)}
```

Comment:
Please convert the `list[TFRepo]` to a plain list of dictionaries; this will avoid unexpected side effects in the future. E.g. https://github.com/app-sre/qontract-reconcile/blob/master/reconcile/skupper_network/integration.py#L307